packages/bierzo-wallet/src/routes/receivePayments/index.tsx

```diff
@@ -0,0 +1,61 @@
+import { Identity } from '@iov/bcp';
+import { TransactionEncoder } from '@iov/encoding';
+import React from 'react';
+import * as ReactRedux from 'react-redux';
+
+import { history } from '..';
+import { getChainName } from '../../config';
+import { getCodecForChainId } from '../../logic/codec';
+import { RootState } from '../../store/reducers';
+import { BwToken } from '../../store/tokens';
+import { BALANCE_ROUTE } from '../paths';
+import Layout, { ChainAddressMap } from './components';
+
+function onReturnToBalance(): void {
+  history.push(BALANCE_ROUTE);
+}
+
+const ReceivePayment = (): JSX.Element => {
+  const tokens = ReactRedux.useSelector((state: RootState) => state.tokens);
+  const pubKeys = ReactRedux.useSelector((state: RootState) => state.extension.keys);
+
+  const [chainAddressMap, setChainAddressMap] = React.useState<ChainAddressMap[]>([]);
+
+  React.useEffect(() => {
+    async function processAddresses(
+      tokens: { [key: string]: BwToken },
+      pubKeys: { [chain: string]: string },
```

Comment:
You don't need the tokens here. You can just use the pubkeys and loop over the identities. An Identity is always a chainId/pubkey pair:

```ts
for (const plainPubkey of Object.values(pubKeys)) {
  const identity: Identity = TransactionEncoder.fromJson(JSON.parse(plainPubkey));
  const address = (await getCodecForChainId(identity.chainId)).identityToAddress(identity);
  const chainName = getChainName(identity.chainId);
}
```